### DISCUSSION TITLE
just adding <div x-data> in <template> makes me idiot..(image uploader component)

### DISCUSSION THREAD

**Discussion by capymind** [alpinejs/alpine#3346] [1. Help] [answered]
Hello, everyone.
First of all, thanks for maintaining this wonderful open source project.

I am a beginner just digging into Alpine.js.
I want to make so-call "image uploader component" which helps user upload images and see thumbnails.
(One of features is to delete image by clicking that image, dispatching delete event and updating rendering.)

The below code works the way I have intended after deleting `x-data` in `<div>` inside `<template>`.
Before deleting that attribute, the order image thumbnails are deleted & added makes me lost, hard to understand the reason.

I think I am not figure out exactly the way `x-data`  works,
Any thought or explanation save me a lot... Thanks in advance.

- codepen: https://codepen.io/kssflow/pen/XWBWyBO

```html
<body>
    <h3>Image Upload & Thumbnail</h3>
    <div>
        <div
            x-data="{
                files: [],
                handleUpload(e) {
                    console.log('handleUpload', $refs.input.files)
                    this.files = $refs.input.files
                },
                handleDelete(index) {
                    console.log('index', index)
                    let fileArray = Array.from(this.files)
                    fileArray.splice(index, 1)
                    let dt = new DataTransfer()
                    fileArray.forEach((file) => dt.items.add(file))
                    console.log(fileArray)
                    this.files = dt.files
                    $refs.input.files = dt.files
                    console.log($refs.input.files)
                }
            }"
            @delete="handleDelete($event.detail.index)"
        >

            <input
                class="d-none"
                name="files"
                type="file"
                accept="image/png, image/jpeg"
                multiple
                x-ref="input"
                @change="handleUpload"
            >

            <button
                class="btn btn-outline-primary"
                type="button"
                @click="$refs.input.click()"
            >
                <i class="bi bi-image"></i>
                <span>Upload Image(.png, .jpg) </span>
            </button>
            <template x-if="files.length > 0">
                <template
                    x-for="(file, index) in files"
                    :key="index"
                >
                    <div> <!--👈😭⚠️solves problem(awkwardly deleted & added image) after deleting 'x-data' ⚠️-->
                        <img
                            :src="URL.createObjectURL(file)"
                            alt="thumbnail"
                            width="150px"
                            height="150px"
                            style="object-fit: cover;"
                            @load="URL.revokeObjectURL($el.src)"
                            @click="$dispatch('delete', { index: index })"
                        >
                        <span x-text="file.name"></span>
                    </div>
                </template>
            </template>
        </div>
    </div>
</body>
```

**Comment by ekwoka** [comment, marked as answer]
x-data just defines some component context.

This issue of you needing the `div` around the `img` and `span` is that `template` elements can have only one child element

this is part of how the template element works.

this is also related to why your `x-if` wrapping the `x-for` doesn't work properly. It won't easily know to remove all the x-for generated elements

You don't actually need to proactively check if the array is not empty, so long as it's an array. The x-for will understand there are no elements and render nothing automatically.

If the condition was more complex for the x-if, I'd do `x-for="file in condition ? files : []"` to short it to an empty array to block rendering.

I'm not sure if you have any other question in there, but this is what I see as issues in your code.

**Reply by capymind** [reply]
Thanks for a kind and helpful comment!
In addition to the original problem, you let me know other code convention and alpine.js directives caveat.

I just got surprised a lot by a huge difference in behavior between `<div>` and `<div x-data>` in my code.
(`<div x-data>` totally destory the functionality that I want to implement...)


https://user-images.githubusercontent.com/108768169/208589581-2f9ae8c0-5545-4f7c-a6f3-4a5e2f9e8508.mp4


https://user-images.githubusercontent.com/108768169/208589660-c4ff09d6-6308-4358-933f-ffd0ccc0a0fa.mp4


Maybe I need more hours involved and expriences about writing right alpine code in alpine way.
Anyway, thanks a lot again.

**Reply by ekwoka** [reply]
I'm not sure why adding an empty x-data would break it...but there is never a context in which you should add an empty x-data inside of an existing component.

I think it's just that an empty x-data makes the current context completely empty, so it was blocking all the context coming down the tree from above.

empty x-data is only meant to be used to allow directives to work that don't depend on a data context (and are thus outside of an actual component) like maybe some that interact with a store, or trigger some other native JS stuff. But you already get all of that if you are in an existing component

Another note: there is no benefit for using dispatch here.

the looped elements have access to the `handleDelete` from the component already. They can call it directly.

```js
@click="handleDelete(index)"
```

$dispatch would be used for components that are siblings to be able to communicate via event on the window. Doesn't make as much sense for within a single component.

**Reply by capymind** [reply]
Thanks a lot @ekwoka 

I think that what you have mentioned about my code is totally right...😭
I have to practice writing many codes with thinking about structure and (re)factoring. 🔥 
(all modern JS frameworks seem quite mysterious for me(beginner) without cautious thinking of internals, if I would be skilled just looking source code makes sense but... I hope some documentation about internals someday)

Thanks for sharing your precious time and efforts. 🙏
